### ISSUE TITLE
Official agreement from code owners required

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#7032]
It was decided yesterday to distribute the merging process (it used to be done by a single person).
The new process is described here: https://github.com/coq/coq/blob/master/dev/doc/MERGING.md
The maintainers are listed here: https://github.com/coq/coq/blob/master/.github/CODEOWNERS

Given that maintainers of different components are now responsible for merging pull requests (after following the process), it is necessary to give push access to more people.

@amahboubi @aspiwack @barras @bgregoir @ejgallego @fajb @forestjulien @gares @herbelin @JasonGross @letouzey @Matafou @mattam82 @maximedenes @ppedrot @silene @SkySkimmer @thery @vbgl @Zimmi48 and **Pierre Corbineau**: you are the listed code owners (did I forget anyone?). Have you read the merging documentation? Do you agree with the process? If yes, please post a comment here saying:

> I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

If not, please say so.

I will need to add all the people who agreed to be code owners as members of the new @coq/pushers team which will give them (if they hadn't already) the right to merge approved pull requests into master. For those who are not yet part of the @coq organization, this will also require accepting the invitation to join the organization on GitHub.

**Comment by Zimmi48** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

Théo Zimmermann

**Comment by thery** [comment]
I agree to be a code owner. I have read the merging documentation. I
will follow the process and use my merging right wisely.

**Comment by ejgallego** [comment]
```coq
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.
```

**Comment by ppedrot** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely. And my credit card number is not 5100 4057 7539 9735.

**Comment by maximedenes** [comment]
> And my credit card number is not 5100 4057 7539 9735.

Just curious, what is not its expiration date?

**Comment by ppedrot** [comment]
@maximedenes Definitely not 42/18.

**Comment by Matafou** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

Could you make the documentation of the merging script a bit more precise please: under which branch should it be launched? A local branch with upstream set to official coq master (or any branch we want to merge in)?

**Comment by bgregoir** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

B.

**Comment by letouzey** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

Pierre

**Comment by ejgallego** [comment]
@Zimmi48 what about signing? Should we require every committer to sign the merge commit?

**Comment by Zimmi48** [comment]
Good question @ejgallego! This is not currently documented but indeed, the merging script automatically signs merge commits (and thus requires gnupg and a PGP key). This is in line with @Matafou's remark that the documentation of the merging script should be made more precise.

**Comment by amahboubi** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

**Comment by JasonGross** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

**Comment by vbgl** [comment]
I agree to be a code owner.
I’ve read the [merging documentation](https://github.com/coq/coq/blob/e554ad813af2e51606284cc628202919ba89d34c/dev/doc/MERGING.md).
I’ll follow the process and use my merging right wisely.

**Comment by herbelin** [comment]
Hum, difficult task to respect the solemn turn taken by this issue without looking too formal either. Let me say the following:

First, I'm not owning any code. There is code I feel responsible of. There is code I have an expertise of. The only reason I see to use the word `owning` in the context of merging is that we are relying on a file conventionally named `CODEOWNERS` by GitHub and of which we cannot change the name. Nevertheless, I gladly accept to have my name listed in front of some components in this file, and, for these components, to try to do reasonably my best to play the role described in the merging documentation which we wrote.

I have a lot to say and already many questions about the process, but let's go by steps and start experimenting in real life with this new protocol! 🍻 to the health of Coq!

**Comment by Zimmi48** [comment]
Yeah I used the wording "code owner" because it is the one of GitHub, definitely not to mean that anyone is owning any code.
About the formal wording that I asked, it was just to make sure that nobody answered just "Yeah, OK" without really reading the doc but your reformulation of this is fine with me too.

**Comment by herbelin** [comment]
No problem, I perfectly understand the reason.

**Comment by silene** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

**Comment by aspiwack** [comment]
Good for me.

**Comment by Zimmi48** [comment]
@aspiwack Have you read the documentation?

**Comment by aspiwack** [comment]
Oh. And just to be clear: the secondary maintainers, from the point of view of Github, are just semanticless comments, right?

**Comment by aspiwack** [comment]
> @aspiwack Have you read the documentation?

Of course.

**Comment by Zimmi48** [comment]
> And just to be clear: the secondary maintainers, from the point of view of Github, are just semanticless comments, right?

Yes.

**Comment by gares** [comment]
I'm in.



> Could you make the documentation of the merging script a bit more precise please: under which branch should it be launched? A local branch with upstream set to official coq master (or any branch we want to merge in)?

@Matafou in #7029 I'm trying to make the script more user friendly (I used it once and I was not happy with it).  I invite everybody to do the same at his turn, including improving MERGING.md if needed)

**Comment by Matafou** [comment]
OK thanks.

**Comment by Matafou** [comment]
With my ownership declared I am receiving lots of PR notification messages now, even ones that do not concern me.

I am considering changing my git notification settings to 
```
* Participating - Notifications for the conversations you are participating in, or if someone cites you with an @mention.
 Email Yes
 Web  Yes

* Watching - Notifications for all repositories, teams, or conversations you’re watching.
 Email: no
 Web: no
```
Will I still receive notifications about the code I "own" (as a secondary owner)?

**Comment by silene** [comment]
> Will I still receive notifications about the code I "own" (as a secondary owner)?

I was about to answer yes, but I now notice that I receive notifications about `doc/` because I am actually the primary owner of the directory. I thought I was only a secondary owner. I don't feel like being the primary owner of a subsystem I didn't agree with (i.e. sphinx).

**Comment by maximedenes** [comment]
@silene Is is ok for you if we switch? I can become the principal maintainer of `doc/`.

**Comment by maximedenes** [comment]
@Matafou For a short period, secondary owners did receive notifications, but it is no longer the case after #7027, precisely because it was generating too much noise.

**Comment by Matafou** [comment]
Mmmh OK we will see.

**Comment by silene** [comment]
> Is is ok for you if we switch? I can become the principal maintainer of `doc/`.

Yes. Thanks.

**Comment by mattam82** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

I'm not so sure to be the right primary for vernac/ but we'll see.

**Comment by ejgallego** [comment]
> I'm not so sure to be the right primary for vernac/ but we'll see.

`vernac` is quite a beast indeed as it agglomerates many different subsystems. Maybe we could refine a bit here but it is not easy.

**Comment by Zimmi48** [comment]
Please people when you say you have read the doc and you will follow the process, do it. There are already pull requests that have been merged without using the merge script. This breaks in turn the backport script. If you want to make an exception to the process in some case, ask before doing it instead of "judging" on your own that it is fine.

**Comment by forestjulien** [comment]
I agree to be a code owner. I have read the merging documentation. I
will follow the process and use my merging right wisely.

**Comment by Zimmi48** [comment]
FTR I wrote an e-mail to Pierre Corbineau to inform him of this.

**Comment by SkySkimmer** [comment]
Is it OK to merge without signing?

**Comment by maximedenes** [comment]
Unless there is a concrete motivation against it, I think it would be better to sign the merge commits.

**Comment by SkySkimmer** [comment]
The motivation is not wanting to set up signing.

**Comment by maximedenes** [comment]
I think it's worth setting up signing, but:
1- we should document it to make it easier
2- if people think it is not important, I won't fight for it :)

**Comment by aspiwack** [comment]
@maximedenes it would probably help if you shared why you think signing merge commits is useful. Then we can weigh the different options.

**Comment by maximedenes** [comment]
Whether it is a good idea to sign commits or not is a standard debate in the git world, with standard reads like: https://mikegerwitz.com/papers/git-horror-story

The motivation is to avoid someone impersonating a maintainer when integrating a change. But I personally sign all my commits, I'm not saying there's something specific to merge commits.

```
git config user.name "Gaëtan Gilbert"'
git config user.email gaetan.gilbert@skyskimmer.net
dev/tools/merge-pr.sh XXX
git push
```

**Comment by aspiwack** [comment]
If you sign all your commit, I doubt there is a need for the merge tool to sign itself, I suppose you've set up your git to sign commits automatically, right?

_after having a look at the script_

In fact the script doesn't seem to use gpg, only to require it as a dependency. Which doesn't seem particularly useful. Am I missing something?

**Comment by maximedenes** [comment]
The tool doesn't merge itself, it asks git to do it.

https://github.com/coq/coq/blob/546a472b298ed00d0fea6de3145141f76ecb576c/dev/tools/merge-pr.sh#L149

**Comment by maximedenes** [comment]
> suppose you've set up your git to sign commits automatically, right?

I have, but I guess @SkySkimmer hasn't, hence his question.

**Comment by aspiwack** [comment]
Ah, but you call `merge -S`, my bad. So it will fail if gpg doesn't work properly.

But how about just calling `git merge` without `-S`? It will still sign your commits but not those of those who don't want to sign.

We may collectively decide to sign commits or merge commits, or tags, or something. But it's a matter of policy which seems to be orthogonal to the issue we're addressing here.

**Comment by maximedenes** [comment]
I'm a bit lost. I thought the entire point of this merge script was to enforce the same policy for all mergers.

**Comment by aspiwack** [comment]
I'm not arguing against this. But since I'm guessing that « commits musn't be signed » is not a policy we're considering enforcing. And since as far as I'm aware, signing commits has never been a policy. It seems to me, that we'd better discuss whether to make it policy in a second, separate step.

As this issue seems to be mostly about the review and assignment process.

**Comment by Zimmi48** [comment]
## Why signing merge commit is more important than requesting our contributors to sign their commits

Because the last commit in master at any time is always a merge commit. Thus, the HEAD of master is always signed.

## Why the script calls `git merge -S`

Because I asked that it does when I started merging. @maximedenes had been signing his commits for at least a year when I started to do backport and merge commits. I viewed it as a regression if I was not signing my own merge commits so I set up signing for my commits. However, I am personally unsure that signing *all* one's commits is a good practice. Personally I want signing to be a conscious action, I set my gpg agent to save my passphrase only for 10 minutes and I keep it long so I don't want to type it all the time. Thus, I do not sign WIP commits. I only sign commits I want integrated. That's why I requested that the merging script be modified to explicitly use the `-S` option.

**Comment by PierreCorbineau** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

**Comment by aspiwack** [comment]
> Because I asked that it does when I started merging.

Yes, sure. Until this week, you and Maxime were the only two users of this script. So this has been tailored to your process. But if we are to generalise its usage, it makes sense to review these decisions collectively.

Regarding signing, it seems that everybody has a very different opinion on about every aspects of it. So it may require a dedicated thread.

**Comment by Zimmi48** [comment]
> Regarding signing, it seems that everybody has a very different opinion on about every aspects of it. So it may require a dedicated thread.

Feel free (you or anyone else) to open a new issue or a PR removing this `-S` and we can discuss there.

**Comment by gares** [comment]
@SkySkimmer and all other that don't know GPG. It has the most masochistic user interface ever, but thanks god git call it for you.  The only thing you need to do is to create a key.  This is a nice tutorial (recommended in Debian, were we use GPG for signing uploads):
  https://ekaia.org/blog/2009/05/10/creating-new-gpgkey/

**Comment by gares** [comment]
FTR, I think we can defend signing commits with better arguments than the ones used to defend the white space linter. Just make a key, it will take way less energy.

**Comment by ejgallego** [comment]
Signing merge commits with the key you are registered in GitHub should be standard policy and any non-signed of wrongly signed commit should be considered coming from a compromised entity.

**Comment by MSoegtropIMC** [comment]
I agree to be a code owner.
I have read the merging documentation.
I will follow the process and use my merging right wisely.

Michael Soegtrop

I have a few questions, though:
- Is the merge script safe to be run from Cygwin, assuming the required tools are there?
- Is there a documented trust chain for the required GPG keys? That is who should sign that a specific key actually belongs to me? What root authorities does the entity which checks this have? The script doesn't seem to check this (it relies on the settings of the local GPG installation, which can be setup in quite a few different ways)
- Are there any rules for handling the private key (basic private key security ... key encryption with password of this and that complexity not stored electronically ... pin protected OpenPGP smart card with pin pad reader)

**Comment by Zimmi48** [comment]
> Is the merge script safe to be run from Cygwin, assuming the required tools are there?

I don't know, I guess testing will tell.

> Is there a documented trust chain for the required GPG keys? That is who should sign that a specific key actually belongs to me? What root authorities does the entity which checks this have?

That's an interesting discussion to have (probably in a dedicated thread). For now, there is no real requirement but it's better if you give your public key to GitHub so that it can displays that the commit was signed with a valid signature from the account owner.

> Are there any rules for handling the private key (basic private key security ... key encryption with password of this and that complexity not stored electronically ... pin protected OpenPGP smart card with pin pad reader)

There are no such rules right now and we don't want to make it too hard to be a code owner so probably no strong requirements here but it would still be very interesting to discuss.

Note that the backporting script that is used to manage the stable branch checks that the backported PR was merged with a trusted signature.

**Comment by SkySkimmer** [comment]
I agree to be a code owner.
I have read the merging documentation.
I will follow the process and use my merging right wisely.

**Comment by Zimmi48** [comment]
We are now 20 code owners. The only people we didn't hear back from yet are @barras and @fajb.

**Comment by fajb** [comment]
I agree to be a code owner. I have read the merging documentation. I will follow the process and use my merging right wisely.

**Comment by Zimmi48** [comment]
Thanks @fajb! I've invited you to the pushers team.

**Comment by Zimmi48** [comment]
I guess we can close this issue now... @barras has his own dedicated issue #7346.
